File: Compiler/src/main/kotlin/nodes/VarNode.java
package nodes;

public class VarNode {
    public VarItemNode first;
    public VarItemNode last;

    public VarNode(VarItemNode item) {
        this.first = item;
        this.last = item;
    }

    public static VarNode createIdVarNode(String id) {
        VarItemNode item = new VarItemNode();
        item.id = Utils.getLastId();
        item.type = VarType.IDENT;
        item.ident = id;
        return new VarNode(item);
    }

    public void addExpr(ExprNode expr) {
        VarItemNode item = new VarItemNode();
        item.id = Utils.getLastId();
        item.type = VarType.VAR;
        item.second_expr = expr;
        this.last.next = item;
        this.last = item;
    }

    public void addId(String id) {
        VarItemNode item = new VarItemNode();
        item.id = Utils.getLastId();
        item.type = VarType.VAR;
        item.ident = id;
        item.is_map_key = true;
        this.last.next = item;
        this.last = item;
    }

    public static VarNode createFunctionWithExprVarNode(ExprNode function_call, ExprNode expr) {
        VarItemNode item = new VarItemNode();
        item.id = Utils.getLastId();
        item.type = VarType.FUNCTION_CALL;
        item.first_expr = function_call;
        item.second_expr = expr;
        return new VarNode(item);
    }

    public static VarNode createFunctionWithIdVarNode(ExprNode function_call, String id) {
        VarItemNode item = new VarItemNode();
        item.id = Utils.getLastId();
        item.type = VarType.FUNCTION_CALL;
        item.first_expr = function_call;
        item.ident = id;
        item.is_map_key = true;
        return new VarNode(item);
    }

    public static VarNode createExprWithExprVarNode(ExprNode expr1, ExprNode expr2) {
        VarItemNode item = new VarItemNode();
        item.id = Utils.getLastId();
        item.type = VarType.ADJUSTED_EXPR;
        item.first_expr = expr1;
        item.second_expr = expr2;
        return new VarNode(item);
    }

    public static VarNode createExprWithIdVarNode(ExprNode expr, String id) {
        VarItemNode item = new VarItemNode();
        item.id = Utils.getLastId();
        item.type = VarType.ADJUSTED_EXPR;
        item.first_expr = expr;
        item.ident = id;
        item.is_map_key = true;
        return new VarNode(item);
    }
}

File: Compiler/src/main/kotlin/nodes/VarNode.kt
package nodes

class VarNode(var first: VarItemNode) {
    var last: VarItemNode = first

    fun add(varItemNode: VarItemNode) {
        last.next = varItemNode
        last = varItemNode
    }

    fun addExpr(expr: ExprNode?) {
        val item = VarItemNode()
        item.id = Utils.lastId
        item.type = VarType.VAR
        item.secondExpr = expr
        last.next = item
        last = item
    }

    fun addId(id: String?) {
        val item = VarItemNode()
        item.id = Utils.lastId
        item.type = VarType.VAR
        item.ident = id!!
        item.isMapKey = true
        last.next = item
        last = item
    }

    companion object {
        fun createIdVarNode(id: String?): VarNode {
            val item = VarItemNode()
            item.id = Utils.lastId
            item.type = VarType.IDENT
            item.ident = id!!
            return VarNode(item)
        }

        fun createFunctionWithExprVarNode(function_call: ExprNode?, expr: ExprNode?): VarNode {
            val item = VarItemNode()
            item.id = Utils.lastId
            item.type = VarType.FUNCTION_CALL
            item.firstExpr = function_call
            item.secondExpr = expr
            return VarNode(item)
        }

        fun createFunctionWithIdVarNode(function_call: ExprNode?, id: String?): VarNode {
            val item = VarItemNode()
            item.id = Utils.lastId
            item.type = VarType.FUNCTION_CALL
            item.firstExpr = function_call
            item.ident = id!!
            item.isMapKey = true
            return VarNode(item)
        }

        fun createExprWithExprVarNode(expr1: ExprNode?, expr2: ExprNode?): VarNode {
            val item = VarItemNode()
            item.id = Utils.lastId
            item.type = VarType.ADJUSTED_EXPR
            item.firstExpr = expr1
            item.secondExpr = expr2
            return VarNode(item)
        }

        fun createExprWithIdVarNode(expr: ExprNode?, id: String?): VarNode {
            val item = VarItemNode()
            item.id = Utils.lastId
            item.type = VarType.ADJUSTED_EXPR
            item.firstExpr = expr
            item.ident = id!!
            item.isMapKey = true
            return VarNode(item)
        }
    }
}